src/error.ts

```diff
@@ -1,51 +1,157 @@
-import * as fs from "fs/promises"
-import * as jsonc from "jsonc-parser"
+import axios from "axios"
+import * as forge from "node-forge"
+import * as tls from "tls"
 import * as vscode from "vscode"
-import { Storage } from "./storage"
 
-export class SelfSignedCertificateError extends Error {
-  public static Notification =
-    "Your Coder deployment is using a self-signed certificate. VS Code uses a version of Electron that does not support registering self-signed intermediate certificates with extensions."
+// X509_ERR_CODE represents error codes as returned from BoringSSL/OpenSSL.
+export enum X509_ERR_CODE {
+  UNABLE_TO_VERIFY_LEAF_SIGNATURE = "UNABLE_TO_VERIFY_LEAF_SIGNATURE",
+  DEPTH_ZERO_SELF_SIGNED_CERT = "DEPTH_ZERO_SELF_SIGNED_CERT",
+  SELF_SIGNED_CERT_IN_CHAIN = "SELF_SIGNED_CERT_IN_CHAIN",
+}
+
+// X509_ERR contains human-friendly versions of TLS errors.
+export enum X509_ERR {
+  PARTIAL_CHAIN = "Your Coder deployment's certificate cannot be verified because a certificate is missing from its chain. To fix this your deployment's administrator should bundle the missing certificates or you can add the missing certificates directly to this system's trust store.",
+  // NON_SIGNING can be removed if BoringSSL is patched and the patch makes it
+  // into the version of Electron used by VS Code.
+  NON_SIGNING = "Your Coder deployment's certificate is not marked as being capable of signing. VS Code uses a version of Electron that does not support certificates like this even if they are self-issued. The certificate should be regenerated with the certificate signing capability.",
+  UNTRUSTED_LEAF = "Your Coder deployment's certificate does not appear to be trusted by this system. The certificate should be added to this system's trust store.",
+  UNTRUSTED_CHAIN = "Your Coder deployment's certificate chain does not appear to be trusted by this system. The root of the certificate chain should be added to this system's trust store. ",
```

Comment:
@bpmct the issue https://github.com/coder/vscode-coder/issues/116 says to link to docs but do you think these instructions are sufficient?  Or do you think we will also need a docs page that goes into more detail, like what exactly it means to bundle, how to serve a bundled certificate, how to add to the trust store to each operating system, how to add the key signing capability, that sort of thing?

Reply:
Alternatively we could offload all the work to docs, basically if we get any TLS error we link to the docs and then in the docs we describe how to figure out what the problem is with your certificate (rather than trying to detect it automatically here) and how to fix it.

Reply:
Merging for now, but happy to make changes later!